File: leetcode/403_Frog_Jump.py
from unittest import TestCase
from typing import List

class Solution:
    def canCross(self, stones: List[int]) -> bool:
        hieghts = set(stones)
        target = stones[-1]
        dic = {}

        def rec(h: int, k: int):
            next_h = h + k
            if next_h == target:
                return True
            if (h, k) not in dic:
                dic[(h, k)] = False
                if next_h in hieghts:
                    for next_k in (k + 1, k, k - 1):
                        if rec(next_h, next_k):
                            dic[(h, k)] = True
                            break
            return dic[(h, k)]

        return rec(0, 1)
            

class CanCrossTest(TestCase):
    def test_1(self):
        self.assertTrue(Solution().canCross([0,1,3,5,6,8,12,17]))

    def test_2(self):
        self.assertFalse(Solution().canCross([0,1,2,3,4,8,9,11]))        

File: leetcode/403_Frog_Jump.rs
use std::collections::HashSet;
struct Solution {}

impl Solution {
    pub fn can_cross(stones: Vec<i32>) -> bool {
        let target = *stones.last().unwrap() as i64;
        let heights = stones.into_iter().map(|v| v as i64).collect::<HashSet<_>>();
        let mut visited = HashSet::new();
        let mut stack = vec![(0, 1)];

        while let Some((h, k))  = stack.pop() {
            let next_h = h + k;
            if next_h == target {
                return true;
            }
            if heights.contains(&next_h) && visited.insert((h,k)) {
                for dk in [-1, 0, 1] {
                    stack.push((next_h, k + dk));
                }
            }
        }
        false
    }
}

#[test]
fn test_1() {
    assert!(Solution::can_cross(vec![0,1,3,5,6,8,12,17]));
}

#[test]
fn test_2() {
    assert!(!Solution::can_cross(vec![0,1,2,3,4,8,9,11]));
}
